### ISSUE TITLE
Cannot use a lambda expression as an argument to a dynamically dispatched operation without first casting it to a delegate or expression tree type.

### ISSUE THREAD

**Issue by axdiamond** [dotnet/roslyn#23221]
**Version Used**: 
 .NET Framework 4.5.2
Tested in both C# 6 and 7, with both VS 2015 and 2017, respectively. 

**Steps to Reproduce**:
Here is the problem code. 

    static void Main()
    {
        dynamic dynamicString = "123";
        var parsedInt = int.Parse(dynamicString);
        TestMethod(parsedInt).Select(o => o);
    }

    static List<int> TestMethod(int id)
    {
        return new List<int>
        {
            1, 2, 3, 4, 5, 6
        };
    }

I am expecting that the return value of `int.Parse` is an int, but the compiler is reporting that it is a dynamic. 

If I change `var parsedInt` to `int parsedInt`, everything compiles normally, and runs properly.

Instead of compiling, it throws 
```Cannot use a lambda expression as an argument to a dynamically dispatched operation without first casting it to a delegate or expression tree type.```

Am I misunderstanding how dynamic works here?

**Comment by HaloFour** [comment]
Using a `dynamic` anywhere in an expression causes the entire expression to become dynamic and the result is always a `dynamic`.

Using `var` the inferred type of `parsedInt` is `dynamic`.  When you change the type to be explicitly `int` you are adding an implicit dynamic cast.

**Comment by HaloFour** [comment]
Also note that the call to `int.Parse` isn't done statically by the compiler.  The compiler emits the type token for `System.Int32` and then looks up and invokes the method by name using the `CSharp.RuntimeBinder`.  Since the method is static and known at compile time the compiler will at least provide static type checking, though.

**Comment by axdiamond** [comment]
Interesting, Do you have resources for learning more about how the compiler handles dynamic? Or the type system in general?

**Comment by HaloFour** [comment]
I don't know of anything comprehensive.  The [C# Language Specification](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#dynamic-binding) probably gets into it a bit but is probably a dry read.  I found [this site](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/using-type-dynamic) which does explain the behavior above:

> The result of most dynamic operations is itself `dynamic`. For example, if you rest the mouse pointer over the use of `testSum` in the following example, IntelliSense displays the type (local variable) **dynamic testSum**.
> ```cs
> dynamic d = 1;
> var testSum = d + 3;
> // Rest the mouse pointer over testSum in the following statement.
> System.Console.WriteLine(testSum);
> ```
> Operations in which the result is not `dynamic` include conversions from `dynamic` to another type, and constructor calls that include arguments of type `dynamic`. For example, the type of `testInstance` in the following declaration is `ExampleClass`, not `dynamic`.
> ```cs
> var testInstance = new ExampleClass(d);
> ```
